src/app.rs

```diff
@@ -696,13 +692,17 @@ where
     DistrT: Distribution,
     IbcT: Ibc,
     GovT: Gov,
-    CustomT::ExecT: Clone + fmt::Debug + PartialEq + JsonSchema + DeserializeOwned + 'static,
+    CustomT::ExecT: Clone + Debug + PartialEq + JsonSchema + DeserializeOwned + 'static,
     CustomT::QueryT: CustomQuery + DeserializeOwned + 'static,
 {
     /// This registers contract code (like uploading wasm bytecode on a chain),
     /// so it can later be used to instantiate a contract.
-    pub fn store_code(&mut self, code: Box<dyn Contract<CustomT::ExecT, CustomT::QueryT>>) -> u64 {
-        self.init_modules(|router, _, _| router.wasm.store_code(code) as u64)
+    pub fn store_code(
+        &mut self,
+        creator: Addr,
+        code: Box<dyn Contract<CustomT::ExecT, CustomT::QueryT>>,
+    ) -> AnyResult<u64> {
```

Comment:
Yes, I agree, generally such a change is like breaking a leg during running. But currently this library has not yet stabilized API and having a possibility to specifiy a creator (like in real blockchain), seems to be ok when testing contracts. Of course, we could do this in few steps, like adding new function, deprecating the old one and then removing the old one, but then we end up with function named like: **store_code_with_creator**. Changing the contract now will break some tests (which should be easy to fix) but we have then a stable contract with all needed attributes I guess.

Reply:
I know we haven't stabilized our API yet, but I think we should still try to avoid breaking unless there is a really good reason to do so. I don't feel like avoiding a somewhat cumbersome function name is a good reason to do that, especially since this change breaks all of our users.
But this is just my opinion, so take it with a grain of salt.
Anyways, I also just wanted to draw attention to this because we need to keep it in mind the next time we release, as we have to bump the minor version if we decide to break.

Reply:
Ok, I agree then. I will provide a function named **store_code_with_creator** that will take an additional argument named **creator** of type _Addr_. Function **store_code** will stay intact. The default creator to be used internally when calling **store_code** will be:
```
Addr::unchecked("code-storer")
```

Reply:
@chipshort Unfortunately, in this case, the breaking change requires long term - the best we can do is to feature flag it, but it would probably be an overhead.

Adding another function with a workaround is a way to go, but it is noisy to me - we can do that temporarily to postpone breaking API. Still, some more breaks may come when we stabilize the usages of MT anyway, so at the end of the day, we will break this function before 1.0.0.

Therefore - even if, for now, we decide not to break the API (I am not a fan of this - it is not a very painful change; you can almost fix it with `sed`), we need to keep some track (maybe GH issue) with API to clean up before 1.0.0. The only reason not to break this API to me now is "because we might want to change it yet another way", and then we might lose track of it. I would even consider adding the feature flag "breaking" where we immediately deliver breaking API (so `store_code` with "breaking" takes an additional argument, and `store_code_with_creator` doesn't exist), which would make it easier for us to keep track of breaking changes and not get back to things done before. Then right before we decide to introduce those breaking changes, we can prepare a release that warns when compiled without a "breaking" feature flag (but not making it default) marks changed APIs as deprecated (with a hint to enable the flag) - this would be purely transitional release. Then upcoming release would completely remove the "breaking" flag and deprecated APIs.

My point in the short version - there are a lot of APIs to be improved in MT especially considering adding new features there, and keeping API non-breaking no matter what would make us end up in the C++ state.

Reply:
I'm not a big fan of the "breaking" feature idea. Cargo features have to be additive. Otherwise, if a transitive dependency enables it, that can break your code.

Reply:
We can provide a feature named **multitest_api_1_0** with proper comment in Cargo.toml and in crate's readme file. This feature should not be enabled until the release of version 1.0.0 (when in fact this feature will be no more needed). But we could also mention, and encourage users to play with this option and to report suggestions to the planned API in version 1.0. Of course in testing the production ready contracts this option should not be used until Multi Test 1.0 is released.

Reply:
@chipshort Is it ok for you?

Reply:
Yes, let's do it like this then